prov/tcp/src/tcpx_rma.c

```diff
@@ -107,7 +107,71 @@ static ssize_t tcpx_rma_readv(struct fid_ep *ep, const struct iovec *iov, void *
 static ssize_t tcpx_rma_writemsg(struct fid_ep *ep, const struct fi_msg_rma *msg,
 		uint64_t flags)
 {
-	return -FI_ENODATA;
+	struct tcpx_ep *tcpx_ep;
+	struct tcpx_cq *tcpx_cq;
+	struct tcpx_pe_entry *send_entry;
+	uint64_t data_len;
+
+	tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+	tcpx_cq = container_of(tcpx_ep->util_ep.tx_cq, struct tcpx_cq,
+			       util_cq);
+
+	send_entry = tcpx_pe_entry_alloc(tcpx_cq);
+	if (!send_entry)
+		return -FI_EAGAIN;
+
+	assert(msg->iov_count <= TCPX_IOV_LIMIT);
+
+	data_len = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+
+	assert(!(flags & FI_INJECT) || (data_len <= TCPX_MAX_INJECT_SZ));
+
+	send_entry->msg_hdr.version = OFI_CTRL_VERSION;
+	send_entry->msg_hdr.op = ofi_op_write;
```

Comment:
The rxm provider ended up adding multiple pools based on the operation type.  Each pool had pre-formatted headers, so that it wasn't necessary to format them on every data transfer call.   We should add something similar for tcpx in order to reduce software overhead.  A lot of the fields in send_entry can be preset.